# Dantzig-Wolfe Decomposition — Optimization Problem 3

**Course:** Complements and Tools of Operations Research (CORO)  
**Institution:** ENSIIE — Université Paris-Saclay  
**Professor:** Mahdi Moeini, Eric Soutil   
**Student:** DIN Sokheng  
**Date:** 31 March 2026

---

## 1. Original Problem

We consider the following binary optimization problem:

$$
\min \; -x_1 - 2x_2 - 3x_3 - 2x_4
$$

$$
\text{s.t.} \quad 3x_1 + x_2 + 2x_3 + x_4 \leq 5
$$

$$
-x_1 + x_2 \leq 1
$$

$$
-x_1 + 2x_2 \leq 2
$$

$$
3x_3 + 2x_4 \leq 4
$$

$$
x_1, x_2, x_3, x_4 \in \{0, 1\}
$$

## 2. Identification of the Block Structure

The constraint matrix has a block-diagonal structure with a single linking constraint. We identify:

**Linking constraint** (the complicating constraint coupling both blocks):

$$
3x_1 + x_2 + 2x_3 + x_4 \leq 5
$$

**Block 1** variables $(x_1, x_2)$:

$$
P_1 = \{(x_1, x_2) \in \{0,1\}^2 : -x_1 + x_2 \leq 1, \; -x_1 + 2x_2 \leq 2\}
$$

**Block 2** variables $(x_3, x_4)$:

$$
P_2 = \{(x_3, x_4) \in \{0,1\}^2 : 3x_3 + 2x_4 \leq 4\}
$$

The objective decomposes as $c_1^\top x_1 + c_2^\top x_2$ with:

$$
c_1 = \begin{pmatrix} -1 \\ -2 \end{pmatrix}, \quad
c_2 = \begin{pmatrix} -3 \\ -2 \end{pmatrix}
$$

The linking coefficients are:

$$
A_1 = \begin{pmatrix} 3 \\ 1 \end{pmatrix}^\top, \quad
A_2 = \begin{pmatrix} 2 \\ 1 \end{pmatrix}^\top, \quad
b_0 = 5
$$

## 3. Dantzig-Wolfe Reformulation

By Minkowski's theorem, since each $P_k$ is a bounded polytope (binary variables), any $x_k \in \text{conv}(P_k)$ can be written as a convex combination of the extreme points $\{v_k^1, \ldots, v_k^{J_k}\}$ of $\text{conv}(P_k)$:

$$
x_k = \sum_{j=1}^{J_k} \lambda_k^j \, v_k^j, \quad \sum_{j=1}^{J_k} \lambda_k^j = 1, \quad \lambda_k^j \geq 0
$$

Substituting into the original problem, we obtain the problem:

$$
\min \sum_{k=1}^{2} \sum_{j=1}^{J_k} \bar{c}_k^j \, \lambda_k^j
$$

$$
\text{s.t.} \quad \sum_{k=1}^{2} \sum_{j=1}^{J_k} \bar{a}_k^j \, \lambda_k^j \leq b_0 \quad (\pi)
$$

$$
\sum_{j=1}^{J_k} \lambda_k^j = 1, \quad k = 1, 2 \quad (\sigma_k)
$$

$$
\lambda_k^j \geq 0
$$

where $\bar{c}_k^j = c_k^\top v_k^j$ and $\bar{a}_k^j = A_k \, v_k^j$.

The number of constraints is only $1 + 2 = 3$ (one linking plus two convexity), but the number of columns can be large. We solve this via **column generation**.

## 4. Enumeration of Extreme Points

Since all variables are binary, the extreme points of each block are simply the feasible binary vectors.

### Block 1

We check all $(x_1, x_2) \in \{0,1\}^2$ against $-x_1 + x_2 \leq 1$ and $-x_1 + 2x_2 \leq 2$:

- $v_1^1 = (0,0)$: $-0+0=0 \leq 1$, $-0+0=0 \leq 2$. Feasible.
- $v_1^2 = (0,1)$: $-0+1=1 \leq 1$, $-0+2=2 \leq 2$. Feasible.
- $v_1^3 = (1,0)$: $-1+0=-1 \leq 1$, $-1+0=-1 \leq 2$. Feasible.
- $v_1^4 = (1,1)$: $-1+1=0 \leq 1$, $-1+2=1 \leq 2$. Feasible.

All four binary vectors are feasible for Block 1.

### Block 2

We check all $(x_3, x_4) \in \{0,1\}^2$ against $3x_3 + 2x_4 \leq 4$:

- $v_2^1 = (0,0)$: $0 \leq 4$. Feasible.
- $v_2^2 = (0,1)$: $2 \leq 4$. Feasible.
- $v_2^3 = (1,0)$: $3 \leq 4$. Feasible.
- $v_2^4 = (1,1)$: $5 > 4$. Infeasible.

Three feasible points for Block 2.

Three feasible points for Block 2.

## 5. Column Data

For each extreme point, we compute the objective coefficient $\bar{c}_k^j = c_k^\top v_k^j$ and the linking coefficient $\bar{a}_k^j = A_k v_k^j$.

## 5. Column Data

For each extreme point, we compute $\bar{c}_k^j = c_k^\top v_k^j$ and $\bar{a}_k^j = A_k v_k^j$.

### Block 1

- $v_1^1 = (0,0)$: $\bar{c} = 0$, $\bar{a} = 0$
- $v_1^2 = (0,1)$: $\bar{c} = -2$, $\bar{a} = 1$
- $v_1^3 = (1,0)$: $\bar{c} = -1$, $\bar{a} = 3$
- $v_1^4 = (1,1)$: $\bar{c} = -3$, $\bar{a} = 4$

### Block 2

- $v_2^1 = (0,0)$: $\bar{c} = 0$, $\bar{a} = 0$
- $v_2^2 = (0,1)$: $\bar{c} = -2$, $\bar{a} = 1$
- $v_2^3 = (1,0)$: $\bar{c} = -3$, $\bar{a} = 2$

## 6. Column Generation Algorithm

We solve the problem iteratively via the restricted problem (RMP). At each iteration:

1. **Solve the RMP** with the current subset of columns to obtain primal solution $\lambda^*$ and dual values $\pi^*$, $\sigma_k^*$.

2. **Solve the pricing subproblem** for each block $k$:

$$
z_k^* = \min_{x_k \in P_k} \left( c_k - \pi^* A_k \right)^\top x_k
$$

3. **Compute the reduced cost** of the best column from block $k$:

$$
\bar{c}_k^* = z_k^* - \sigma_k^*
$$

4. If $\bar{c}_k^* < 0$, add the corresponding extreme point as a new column to the RMP. If $\bar{c}_k^* \geq 0$ for all $k$, the current RMP solution is optimal for the full master problem.

### Initialization

We initialize the RMP with the trivial columns $v_1^1 = (0,0)$ and $v_2^1 = (0,0)$, which give a feasible solution $\lambda_1^1 = 1$, $\lambda_2^1 = 1$ with objective value $0$.

In [1]:
using JuMP, GLPK, LinearAlgebra

# --- Enumerate extreme points ---

function enumerate_block1()
    pts = []
    for x1 in 0:1, x2 in 0:1
        if -x1 + x2 <= 1 && -x1 + 2x2 <= 2
            push!(pts, [x1, x2])
        end
    end
    return pts
end

function enumerate_block2()
    pts = []
    for x3 in 0:1, x4 in 0:1
        if 3x3 + 2x4 <= 4
            push!(pts, [x3, x4])
        end
    end
    return pts
end

P1 = enumerate_block1()
P2 = enumerate_block2()

c1 = [-1, -2]
c2 = [-3, -2]
A1 = [3, 1]
A2 = [2, 1]
b0 = 5

cbar1 = [dot(c1, v) for v in P1]
abar1 = [dot(A1, v) for v in P1]
cbar2 = [dot(c2, v) for v in P2]
abar2 = [dot(A2, v) for v in P2]

println("Block 1 extreme points:")
for (i, v) in enumerate(P1)
    println("  v1[$i] = $v,  c_bar = $(cbar1[i]),  a_bar = $(abar1[i])")
end
println("\nBlock 2 extreme points:")
for (i, v) in enumerate(P2)
    println("  v2[$i] = $v,  c_bar = $(cbar2[i]),  a_bar = $(abar2[i])")
end

Block 1 extreme points:
  v1[1] = [0, 0],  c_bar = 0,  a_bar = 0
  v1[2] = [0, 1],  c_bar = -2,  a_bar = 1
  v1[3] = [1, 0],  c_bar = -1,  a_bar = 3
  v1[4] = [1, 1],  c_bar = -3,  a_bar = 4

Block 2 extreme points:
  v2[1] = [0, 0],  c_bar = 0,  a_bar = 0
  v2[2] = [0, 1],  c_bar = -2,  a_bar = 1
  v2[3] = [1, 0],  c_bar = -3,  a_bar = 2


## 7. Iteration 1

### Initial RMP

We start with columns $v_1^1 = (0,0)$ and $v_2^1 = (0,0)$ only.

$$
\min \; 0 \cdot \lambda_1^1 + 0 \cdot \lambda_2^1
$$

$$
\text{s.t.} \quad 0 \cdot \lambda_1^1 + 0 \cdot \lambda_2^1 \leq 5
$$

$$
\lambda_1^1 = 1, \quad \lambda_2^1 = 1, \quad \lambda_k^j \geq 0
$$

The unique solution is $\lambda_1^1 = 1$, $\lambda_2^1 = 1$, with objective $z_{RMP} = 0$.

The linking constraint has slack $5$, so $\pi^* = 0$. Both convexity constraints are satisfied by a single variable at cost $0$, so $\sigma_1^* = 0$, $\sigma_2^* = 0$.

### Pricing subproblem

With $\pi^* = 0$, the modified cost equals the original cost $c_k$ for both blocks.

**Block 1:** $\min_{v \in P_1} (-1, -2)^\top v$. The minimum is $-3$ at $v_1^4 = (1,1)$. Reduced cost: $-3 - 0 = -3 < 0$.

**Block 2:** $\min_{v \in P_2} (-3, -2)^\top v$. The minimum is $-3$ at $v_2^3 = (1,0)$. Reduced cost: $-3 - 0 = -3 < 0$.

Both reduced costs are negative, so we add $v_1^4$ and $v_2^3$ to the RMP.

In [2]:
# --- Column Generation Loop ---
active1 = [1]   # indices into P1; start with v1^1 = (0,0)
active2 = [1]   # indices into P2; start with v2^1 = (0,0)

for iter in 1:3
    println("*" ^ 60)
    println("ITERATION $iter")
    println("*" ^ 60)

    # Solve RMP
    rmp = Model(GLPK.Optimizer)
    set_silent(rmp)

    n1 = length(active1)
    n2 = length(active2)

    @variable(rmp, lam1[1:n1] >= 0)
    @variable(rmp, lam2[1:n2] >= 0)

    @objective(rmp, Min,
        sum(cbar1[active1[j]] * lam1[j] for j in 1:n1) +
        sum(cbar2[active2[j]] * lam2[j] for j in 1:n2)
    )

    @constraint(rmp, linking,
        sum(abar1[active1[j]] * lam1[j] for j in 1:n1) +
        sum(abar2[active2[j]] * lam2[j] for j in 1:n2) <= b0
    )

    @constraint(rmp, conv1, sum(lam1[j] for j in 1:n1) == 1)
    @constraint(rmp, conv2, sum(lam2[j] for j in 1:n2) == 1)

    optimize!(rmp)

    z_rmp = objective_value(rmp)
    println("\nRMP objective: $z_rmp")

    for j in 1:n1
        println("  lam1[$(active1[j])] (v=$(P1[active1[j]])) = $(value(lam1[j]))")
    end
    for j in 1:n2
        println("  lam2[$(active2[j])] (v=$(P2[active2[j]])) = $(value(lam2[j]))")
    end

    pi_val = dual(linking)
    sig1 = dual(conv1)
    sig2 = dual(conv2)
    println("\nDual values: pi = $pi_val, sigma1 = $sig1, sigma2 = $sig2")

    # Pricing subproblem
    mc1 = c1 .- pi_val .* A1
    mc2 = c2 .- pi_val .* A2

    println("\nPricing Block 1: modified cost = $mc1")
    best_rc1, best_idx1 = Inf, -1
    for (i, v) in enumerate(P1)
        val = dot(mc1, v)
        rc = val - sig1
        println("  v1[$i] = $v: obj = $val, reduced cost = $rc")
        if rc < best_rc1
            best_rc1, best_idx1 = rc, i
        end
    end

    println("\nPricing Block 2: modified cost = $mc2")
    best_rc2, best_idx2 = Inf, -1
    for (i, v) in enumerate(P2)
        val = dot(mc2, v)
        rc = val - sig2
        println("  v2[$i] = $v: obj = $val, reduced cost = $rc")
        if rc < best_rc2
            best_rc2, best_idx2 = rc, i
        end
    end

    # Add columns or stop
    added = false
    if best_rc1 < -1e-8 && !(best_idx1 in active1)
        push!(active1, best_idx1)
        println("\n--> Add column v1[$best_idx1] = $(P1[best_idx1])")
        added = true
    else
        println("\nBlock 1: no improving column (rc = $best_rc1)")
    end
    if best_rc2 < -1e-8 && !(best_idx2 in active2)
        push!(active2, best_idx2)
        println("--> Add column v2[$best_idx2] = $(P2[best_idx2])")
        added = true
    else
        println("Block 2: no improving column (rc = $best_rc2)")
    end

    if !added
        println("\n*** All reduced costs >= 0 ***")
        x1_val = sum(value(lam1[j]) * P1[active1[j]][1] for j in 1:n1)
        x2_val = sum(value(lam1[j]) * P1[active1[j]][2] for j in 1:n1)
        x3_val = sum(value(lam2[j]) * P2[active2[j]][1] for j in 1:n2)
        x4_val = sum(value(lam2[j]) * P2[active2[j]][2] for j in 1:n2)
        println("\nRecovered solution: x = ($x1_val, $x2_val, $x3_val, $x4_val)")
        println("DW LP bound: z* = $z_rmp")
        break
    end

    println()
end

************************************************************
ITERATION 1
************************************************************

RMP objective: 0.0
  lam1[1] (v=[0, 0]) = 1.0
  lam2[1] (v=[0, 0]) = 1.0

Dual values: pi = 0.0, sigma1 = 0.0, sigma2 = 0.0

Pricing Block 1: modified cost = [-1.0, -2.0]
  v1[1] = [0, 0]: obj = 0.0, reduced cost = 0.0
  v1[2] = [0, 1]: obj = -2.0, reduced cost = -2.0
  v1[3] = [1, 0]: obj = -1.0, reduced cost = -1.0
  v1[4] = [1, 1]: obj = -3.0, reduced cost = -3.0

Pricing Block 2: modified cost = [-3.0, -2.0]
  v2[1] = [0, 0]: obj = 0.0, reduced cost = 0.0
  v2[2] = [0, 1]: obj = -2.0, reduced cost = -2.0
  v2[3] = [1, 0]: obj = -3.0, reduced cost = -3.0

--> Add column v1[4] = [1, 1]
--> Add column v2[3] = [1, 0]

************************************************************
ITERATION 2
************************************************************

RMP objective: -5.25
  lam1[1] (v=[0, 0]) = 0.25
  lam1[4] (v=[1, 1]) = 0.75
  lam2[1] (v=[0, 0]) = 0.0


## 8. Iteration 2

### Updated RMP

Active columns: $v_1^1 = (0,0)$, $v_1^4 = (1,1)$ for Block 1; $v_2^1 = (0,0)$, $v_2^3 = (1,0)$ for Block 2.

$$
\min \; 0 \cdot \lambda_1^1 - 3\lambda_1^4 + 0 \cdot \lambda_2^1 - 3\lambda_2^3
$$

$$
\text{s.t.} \quad 4\lambda_1^4 + 2\lambda_2^3 \leq 5
$$

$$
\lambda_1^1 + \lambda_1^4 = 1, \quad \lambda_2^1 + \lambda_2^3 = 1
$$

Setting $\lambda_2^3 = 1$ (cost $-3$, linking usage $2$) leaves capacity $3$ for Block 1. With $4\lambda_1^4 \leq 3$, we get $\lambda_1^4 = 3/4$, $\lambda_1^1 = 1/4$.

**RMP solution:** $\lambda_1^1 = 1/4$, $\lambda_1^4 = 3/4$, $\lambda_2^1 = 0$, $\lambda_2^3 = 1$, with $z_{RMP} = -9/4 - 3 = -21/4 = -5.25$.

**Dual values** (from optimality conditions on basic variables):

$$
\lambda_1^1: \quad 0 - \pi \cdot 0 - \sigma_1 = 0 \implies \sigma_1 = 0
$$

$$
\lambda_1^4: \quad -3 - 4\pi - \sigma_1 = 0 \implies \pi = -3/4
$$

$$
\lambda_2^3: \quad -3 - 2\pi - \sigma_2 = 0 \implies \sigma_2 = -3/2
$$

### Pricing subproblem

Modified costs with $\pi^* = -3/4$:

**Block 1:** $(c_1 - \pi^* A_1) = (-1 + 9/4, \; -2 + 3/4) = (5/4, \; -5/4)$.

Evaluating over $P_1$: minimum is $-5/4$ at $v_1^2 = (0,1)$. Reduced cost: $-5/4 - 0 = -5/4 < 0$. Add $v_1^2$.

**Block 2:** $(c_2 - \pi^* A_2) = (-3 + 3/2, \; -2 + 3/4) = (-3/2, \; -5/4)$.

Evaluating over $P_2$: minimum is $-3/2$ at $v_2^3 = (1,0)$. Reduced cost: $-3/2 - (-3/2) = 0$. No improving column.

## 9. Iteration 3

### Updated RMP

Active columns for Block 1: $v_1^1 = (0,0)$, $v_1^4 = (1,1)$, $v_1^2 = (0,1)$. Block 2 unchanged.

The RMP now has enough columns for the LP to find the optimum. Setting $\lambda_2^3 = 1$ (linking usage $2$, remaining capacity $3$), we optimize over Block 1:

$$
\min \; -3\lambda_1^4 - 2\lambda_1^2 \quad \text{s.t.} \quad 4\lambda_1^4 + \lambda_1^2 \leq 3, \quad \lambda_1^1 + \lambda_1^4 + \lambda_1^2 = 1
$$

The intersection of $4\lambda_1^4 + \lambda_1^2 = 3$ and $\lambda_1^4 + \lambda_1^2 = 1$ gives $\lambda_1^4 = 2/3$, $\lambda_1^2 = 1/3$, $\lambda_1^1 = 0$.

**RMP solution:** $z_{RMP} = -3(2/3) - 2(1/3) - 3 = -2 - 2/3 - 3 = -17/3 \approx -5.667$.

**Dual values:**

$$
\lambda_1^2: \quad -2 - \pi - \sigma_1 = 0
$$

$$
\lambda_1^4: \quad -3 - 4\pi - \sigma_1 = 0
$$

Subtracting: $-1 - 3\pi = 0$, so $\pi^* = -1/3$, $\sigma_1^* = -5/3$, $\sigma_2^* = -7/3$.

### Pricing subproblem

**Block 1:** $(c_1 - \pi^* A_1) = (0, \; -5/3)$. Minimum over $P_1$ is $-5/3$. Reduced cost: $-5/3 - (-5/3) = 0$.

**Block 2:** $(c_2 - \pi^* A_2) = (-7/3, \; -5/3)$. Minimum over $P_2$ is $-7/3$. Reduced cost: $-7/3 - (-7/3) = 0$.

All reduced costs are non-negative. The algorithm terminates. The current RMP solution is optimal for the full master problem.

## 10. Recovered Solution and Analysis

From the optimal $\lambda$ values, we recover the original variables:

$$
x_1 = \frac{2}{3}(1) + \frac{1}{3}(0) = \frac{2}{3}, \quad x_2 = \frac{2}{3}(1) + \frac{1}{3}(1) = 1
$$

$$
x_3 = 1 \cdot 1 = 1, \quad x_4 = 1 \cdot 0 = 0
$$

The DW LP relaxation optimal value is:

$$
z_{DW}^* = -\frac{17}{3} \approx -5.667
$$

Since $x_1 = 2/3$ is fractional, this is not an integer solution. To obtain the integer optimum, one would need to apply Branch-and-Price. However, the assignment requires only 3 iterations of the DW algorithm.

## 11. Verification

We verify by solving the problem directly: the LP relaxation (continuous $x \in [0,1]$) and the ILP (binary $x \in \{0,1\}$).

In [3]:
# --- Direct LP relaxation ---
m_lp = Model(GLPK.Optimizer)
set_silent(m_lp)
@variable(m_lp, 0 <= x[1:4] <= 1)
@objective(m_lp, Min, -x[1] - 2x[2] - 3x[3] - 2x[4])
@constraint(m_lp, 3x[1] + x[2] + 2x[3] + x[4] <= 5)
@constraint(m_lp, -x[1] + x[2] <= 1)
@constraint(m_lp, -x[1] + 2x[2] <= 2)
@constraint(m_lp, 3x[3] + 2x[4] <= 4)
optimize!(m_lp)
println("LP relaxation: z = ", objective_value(m_lp))
println("x = ", [value(x[i]) for i in 1:4])

# --- Direct ILP ---
m_ip = Model(GLPK.Optimizer)
set_silent(m_ip)
@variable(m_ip, y[1:4], Bin)
@objective(m_ip, Min, -y[1] - 2y[2] - 3y[3] - 2y[4])
@constraint(m_ip, 3y[1] + y[2] + 2y[3] + y[4] <= 5)
@constraint(m_ip, -y[1] + y[2] <= 1)
@constraint(m_ip, -y[1] + 2y[2] <= 2)
@constraint(m_ip, 3y[3] + 2y[4] <= 4)
optimize!(m_ip)
println("\nILP optimal: z = ", objective_value(m_ip))
println("x = ", [value(y[i]) for i in 1:4])

LP relaxation: z = -6.555555555555555
x = [0.5555555555555556, 1.0, 0.6666666666666666, 1.0]

ILP optimal: z = -5.0
x = [0.0, 1.0, 1.0, 0.0]


## 12. Comparison of Bounds

The three bounds satisfy the theoretical ordering:

$$
z_{IP}^* = -5 \;\geq\; z_{DW}^* = -\frac{17}{3} \approx -5.667 \;\geq\; z_{LP}^* \approx -6.556
$$

This confirms the key property of Dantzig-Wolfe decomposition: the DW LP bound is at least as tight as the standard LP relaxation bound. This occurs because the DW master problem implicitly optimizes over the convex hull of integer-feasible extreme points of each block, which is a tighter relaxation than simply replacing $x \in \{0,1\}$ with $x \in [0,1]$.

The integer optimum is $x^* = (0, 1, 1, 0)$ with $z_{IP}^* = -5$. The integrality gap from the DW bound is $|-5 - (-17/3)| = 2/3$, which is smaller than the gap from the LP relaxation.